### ISSUE TITLE
IETF directories not updating

### ISSUE THREAD

**Issue by samans** [YangModels/yang#1562]
https://github.com/YangModels/yang/tree/main/standard/ietf
The RFC directory has not been updated since 23 November 2023
https://github.com/YangModels/yang/commit/121ac333977486a883fd039afae75db386d99623 is the last commit.

The internet-draft extracted yang modules directory https://github.com/YangModels/yang/tree/main/experimental/ietf-extracted-YANG-modules hasn't been upaded since 20 December 2023 https://github.com/YangModels/yang/commit/ead6cbfe0f003ae20ea8cbc903738be6f051a6bc is the last commit.

The modules are available at the https://www.ietf.org/ietf-ftp/yang/, just not in git.

Can someone please look at this and advise?

**Comment by xorrkaz** [comment]
Yes, this has been broken since the move to the new AWS instance.  I'll admit I do not know why as yet.  Modules were being checked against that git repo, it's just for some reason the push isn't happening correctly.  I should have some time at the IETF 120 meeting to take a deeper look at this.  I might have to break out a big hammer to get it working again.

**Comment by rwestphal** [comment]
> The modules are available at the https://www.ietf.org/ietf-ftp/yang/, just not in git.

Indeed, but apparently not all modules.

*draft-ietf-bier-bier-yang-09* was released a few days ago, but I can't find *ietf-bier.yang* in https://www.ietf.org/ietf-ftp/yang/draftmod/

I realize this is a separate problem though.

**Comment by samans** [comment]
Thanks for all the efforts to get the I-D yang models automatically extracted and put in experimental/ietf-extracted-YANG-modules.  However, there are still (at least one) that is missing.  https://datatracker.ietf.org/doc/draft-ietf-netmod-rfc6991-bis/ was updated and contains two yang modules.  But only one of them was extracted and added to experimental/ietf-extracted-YANG-modules.  This one was added..  ietf-yang-types@2024-08-07.yang, the ietf-inet-types@2024-08-07.yang was not extracted and added to the repository.

**Comment by xorrkaz** [comment]
The problem is this line in the ietf-inet-types module:

`+ '\[[a-zA-Z0-9!"#$%&'+"'"+'()*+,./:;<=>?@\^_`{|}~-]+\])';`

That single " quote character there is causing an imbalance in xym's processing of this module, and thus it is not being extracted.  If you adding a second " to the character class, it would extract, but then you have something weird in there.

@einarnn Since you maintain xym, I wonder if you could review this approach to solving the problem:

`xym.py` line 539:

```diff
-                if line.count("\"") % 2 == 0:
+                if line.count("\"") % 2 == 0 or re.search(r"'[^']+\"", line):
```

Essentially, ignore a double-quote if itself is quoted by a single quote on the same line.  This feels naive, but does work in this case.

**Comment by xorrkaz** [comment]
@einarnn Thoughts on the above?  Do you want me to open an issue with `xym`?

**Comment by einarnn** [comment]
Joe, thanks for raising this again. I actually tried to find the referred content that causes the ***@***.*** to evaluate the suggested fix, but I couldn’t find it where the issue thread suggested. Can you provide a link to the problem text that I can review in conjunction with suggested fix?

**Comment by xorrkaz** [comment]
Sure.  The most unambiguous place to find this is by opening the [XML](https://www.ietf.org/archive/id/draft-ietf-netmod-rfc6991-bis-16.xml) and looking at line 1622.  That is the offending line preventing the module `ietf-inet-types@2024-08-07.yang` from being extracted.

**Comment by einarnn** [comment]
@xorrkaz, I agree that this is not seen to fail. But it leads me to wonder if we can just stop trying to do it this way. If we are extracting from the XML variant of the drafts, can't we depend more explicitly on extracting the `sourcecode element

```xml
<sourcecode><![CDATA[
<CODE BEGINS> file "ietf-inet-types@2024-08-07.yang"
module ietf-inet-types {

...rest of the yang model text...

<CODE ENDS>
]]></sourcecode>
```

The current iteration of the tool was implemented to extract text from a text version of the draft, not the XML version. If we instead depend on matching this regex to the contents of `sourcecode` elements:

```
r"""<CODE BEGINS> file "(.+)"\n(.+)\n<CODE ENDS>"""
```

I may not have that completely right, but the intent is that group 1 will contain YANG file name and group 2 will contain the YANG module.

This style of parsing could be triggered with a new option, maybe `--xml | -x`, leaving the current text file targeted approach as it is.

Thoughts?

**Comment by xorrkaz** [comment]
I think that's a _great_ idea given I believe XML is now mandatory.  And if you throw lxml or maybe xmltodict in here, you could just literally grab the element.

**Comment by xorrkaz** [comment]
Okay.  So I found xym to be a bit broken.  I added the pure regex+XML approach and did a PR.

**Comment by xorrkaz** [comment]
I have implemented this XML support in the yc.o extraction code.  I'll be rolling it out shortly, so I'm hoping we see this extraction problem go away.

**Comment by xorrkaz** [comment]
@samans With the new XML parsing code in place, I see `ietf-inet-types@2024-10-21.yang` properly extracted and in my local repo.  I'm going to kick off a PR for @einarnn to merge, and this thing should then be resolved.

**Comment by samans** [comment]
seems good to me now.  Thanks for fixing this!
